### PR TITLE
Polyfill reportValidation calls and bubble UI for webkit.

### PR DESCRIPTION
**Pull request by mkhatib**
Test this out on any webkit-based browser at [the demo here](https://amp-mk-1.herokuapp.com/examples.build/forms.amp.max.html).

This polyfill the bubble UI for validation reporting in webkit browsers. This matches browsers implementation of one invalid bubble at a time, focusing the invalid input, hide bubble on blur and continuous revalidation as the user types in an active invalid reported input.

- [x] Figure out why the example is working on Safari Desktop and iOS simulator but not on real iOS - This was due to the demo page being served on http instead of https.
- [x] Add and fix tests

ITI: #3343

### CODE REVIEW

**Comment by mkhatib** [comment]
Test this out on any webkit-based browser at [the demo here](https://amp-mk-1.herokuapp.com/examples.build/forms.amp.max.html).

**Comment by mkhatib** [comment]
Want to get some feedback on this before writing tests. Let me know if this looks good!

**Comment by dvoytenko** [comment]
@mkhatib Can't get demo to work.

File: extensions/amp-form/0.1/amp-form.css [hunk 1]
```diff
@@ -18,3 +18,28 @@ form.amp-form-submit-success [submit-success],
 form.amp-form-submit-error [submit-error] {
     display: block;
 }
+
+
+.-amp-validation-bubble {
+    background-color: white;
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.css]: Is this a 3-space indent?

File: extensions/amp-form/0.1/amp-form.css [hunk 2]
```diff
@@ -18,3 +18,28 @@ form.amp-form-submit-success [submit-success],
...
+    box-shadow: 0 5px 15px 0 rgba(0, 0, 0, .5);
+    max-width: 200px;
+    position: absolute;
+    top: -999px;
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.css]: Are they `display:none` by default?

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.css]: Not really. I need to be able to measure their size that's why I put them off the screen. Should I handle this differently?

File: extensions/amp-form/0.1/amp-form.js [hunk 3]
```diff
@@ -177,8 +185,69 @@ export class AmpForm {
 
 
 /**
+ * Reports validity for the first invalid input - if any.
+ * @param {!HTMLFormElement} form
+ * @private
+ */
+function reportFormValidity(form) {
+  const inputs = toArray(form.querySelectorAll('input,select,textarea'));
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: We need to move this to `dom.js` and implement via `:scope` when available.

File: extensions/amp-form/0.1/amp-form.css (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.css]: 4. Just noticed my editor has 4-space indent for css, will fix that to 2 instead.

File: extensions/amp-form/0.1/amp-form.js [hunk 4]
```diff
@@ -177,8 +185,69 @@ export class AmpForm {
...
+  for (let i = 0; i < inputs.length; i++) {
+    if (!inputs[i].checkValidity()) {
+      return reportInputValidity(inputs[i]);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: How do we only do this to only limit to the first input?

File: extensions/amp-form/0.1/amp-form.js [hunk 5]
```diff
@@ -177,8 +185,69 @@ export class AmpForm {
...
+}
+
+
+/**
+ * Focuses and reports the invalid message of the input in a message bubble.
+ * @param {!HTMLInputElement} input
+ * @private
+ */
+function reportInputValidity(input) {
+  input./*REVIEW*/focus();
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: REVIEW -> OK

File: extensions/amp-form/0.1/validation-bubble.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    win.document.body.appendChild(this.bubbleElement_);
+  }
+
+  /**
+   * Hides the bubble off screen.
+   */
+  hide() {
+    this.vsync_.mutate(() => {
+      setStyles(this.bubbleElement_, {
+        top: EDGE,
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/validation-bubble.js]: Why not just hide via `visibility: hidden`?

File: extensions/amp-form/0.1/validation-bubble.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    });
+  }
+
+  /**
+   * Shows the bubble targeted to an element with the passed message.
+   * @param {!HTMLElement} targetElement
+   * @param {string} message
+   */
+  show(targetElement, message) {
+    this.bubbleElement_.textContent = message;
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/validation-bubble.js]: This should be done in mutate normally as well.

File: extensions/amp-form/0.1/amp-form.css (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.css]: Commented further, but probably `visibility: hidden` is a good idea.

File: extensions/amp-form/0.1/amp-form.js [hunk 8]
```diff
@@ -37,6 +38,9 @@ const FormState_ = {
   SUBMIT_SUCCESS: 'submit-success',
 };
 
+/** @type {?./validation-bubble.ValidationBubble} */
+let validationBubble;
```
Comment (erwinmombay) [inline extensions/amp-form/0.1/amp-form.js]: add undefined to allowed type if we dont init with null

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: But in vsync.mutate...

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: sgtm thanks 👍

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: We return from the function.

File: extensions/amp-form/0.1/validation-bubble.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+    let targetRect;
+    let bubbleRect;
+    this.vsync_.run({
+      measure: () => {
```
Comment (jridgewell) [inline extensions/amp-form/0.1/validation-bubble.js]: Please use the [`state` pattern](https://github.com/ampproject/amphtml/issues/3718).

File: extensions/amp-form/0.1/validation-bubble.js (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Yeah I think that would work, this would still give the correct dimensions of the element right? Need to do the pointer events setting as well right?

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: But returns what? The function doesn't declare a return type and doesn't return anything after the for-cycle.

File: extensions/amp-form/0.1/validation-bubble.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/validation-bubble.js]: Yes

File: extensions/amp-form/0.1/validation-bubble.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    this.bubbleElement_ = win.document.createElement('div');
+    this.bubbleElement_.classList.add('-amp-validation-bubble');
+    win.document.body.appendChild(this.bubbleElement_);
+  }
+
+  /**
+   * Hides the bubble off screen.
+   */
+  hide() {
+    this.vsync_.mutate(() => {
```
Comment (jridgewell) [inline extensions/amp-form/0.1/validation-bubble.js]: Ditto.

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: I can just put return on a separate statement and return after the if-clause as well. Or just break from the loop without return?

File: extensions/amp-form/0.1/validation-bubble.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+   * Computes the left position of the bubble.
+   * @param {!../../../src/layout-rect.LayoutRectDef} bubbleRect
+   * @param {!../../../src/layout-rect.LayoutRectDef} targetRect
+   * @return {number}
+   * @private
+   */
+  computeLeft_(bubbleRect, targetRect) {
+    return (targetRect.left -
+        // Center the bubble relative to the target element.
+        (bubbleRect.width - targetRect.width) / 2);
```
Comment (jridgewell) [inline extensions/amp-form/0.1/validation-bubble.js]: We can avoid the `bubbleRect` measurement by using CSS transforms:

```css
.bubble {
  transform: translate(-50%, -100%);
}
```

```js
computeLeft(targetRect) {
  return targetRect.left + targetRect.width / 2;
}
computeTop(targetRect) {
  return targetRect.top - 10;
}
```

File: extensions/amp-form/0.1/amp-form.js [hunk 12]
```diff
@@ -94,12 +98,16 @@ export class AmpForm {
...
   handleSubmit_(e) {
-    if (e.defaultPrevented) {
+    if (this.state_ == FormState_.SUBMITTING) {
+      e.preventDefault();
       return;
     }
 
-    if (this.state_ == FormState_.SUBMITTING) {
+    const shouldValidate = !this.form_.hasAttribute('novalidate');
+    const isInvalid = this.form_.checkValidity && !this.form_.checkValidity();
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: I don't think we should `#checkValidity` if `novalidate` is set.

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Is there a need to `toArray`?

File: extensions/amp-form/0.1/amp-form.js [hunk 13]
```diff
@@ -177,8 +185,69 @@ export class AmpForm {
 
 
 /**
+ * Reports validity for the first invalid input - if any.
+ * @param {!HTMLFormElement} form
+ * @private
+ */
+function reportFormValidity(form) {
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Naming here is odd, since we're actually reporting invalidity.

File: extensions/amp-form/0.1/amp-form.js [hunk 14]
```diff
@@ -177,8 +185,69 @@ export class AmpForm {
...
+  event.target.removeEventListener('keyup', onInvalidInputKeyUp_);
+}
+
+
+/**
+ * Focuses and reports the invalid message of the input in a message bubble.
+ * @param {!HTMLInputElement} input
+ * @private
+ */
+function reportInputValidity(input) {
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Ditto naming.

File: extensions/amp-form/0.1/amp-form.css (hunk 2 above)
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.css]: Using [transforms](https://github.com/ampproject/amphtml/pull/3734#discussion_r68319778), we can use `display: none`.

File: extensions/amp-form/0.1/amp-form.js (hunk 12 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: oh shit. yeah. correct. Will fix.

File: extensions/amp-form/0.1/amp-form.js (hunk 13 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: The native API is called [reportValidity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reportValidity).

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: No need. I had it earlier because I was doing `forEach` but no longer.

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Talk to Dima, will be breaking from the loop instead.

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: @dvoytenko for moving this to `dom.js` When `:scope` is not available do we jsut do `querySelectorAll` or did you want us to navigate the DOM and check child by child (and its children and so on)?

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done. I've wrapped the original call of `reportFormValidation` with the `vsync.mutate` call. Is that alright? Otherwise I'd have to pass down the `window` or the vsync service instance to these functions.

File: extensions/amp-form/0.1/amp-form.css (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.css]: Done.

File: extensions/amp-form/0.1/amp-form.css (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.css]: Sticked with @jridgewell suggestion to use transforms and display none! Thanks!

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done.

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done. I've wrapped the original call of reportFormValidation with the vsync.mutate call. Is that alright? Otherwise I'd have to pass down the window or the vsync service instance to these functions.

File: extensions/amp-form/0.1/validation-bubble.js (hunk 7 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Done.

File: extensions/amp-form/0.1/amp-form.js (hunk 8 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/validation-bubble.js (hunk 9 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Done! Thanks for the detailed explanation in the bug!

File: extensions/amp-form/0.1/validation-bubble.js (hunk 10 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Done.

File: extensions/amp-form/0.1/validation-bubble.js (hunk 11 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Done! Thanks for the tip! This is way better.

File: extensions/amp-form/0.1/amp-form.js (hunk 12 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done.

File: extensions/amp-form/0.1/amp-form.js (hunk 13 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Ack.

File: extensions/amp-form/0.1/amp-form.js (hunk 14 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Ack.

**Comment by mkhatib** [comment]
Updated! Thanks for all the great comments! PTAL.

File: extensions/amp-form/0.1/amp-form.js [hunk 15]
```diff
@@ -94,12 +102,16 @@ export class AmpForm {
...
+      e.preventDefault();
       return;
     }
 
-    if (this.state_ == FormState_.SUBMITTING) {
+    const shouldValidate = !this.form_.hasAttribute('novalidate');
+    if (shouldValidate &&
+        this.form_.checkValidity && !this.form_.checkValidity()) {
       e.preventDefault();
+      this.vsync_.run({mutate: reportValidity}, {form: this.form_});
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: You'll need to use the same `{measure: undefined, mutate: reportValidity}` hidden class. 😞

I'll file an issue to add `state` to `#measure` and `#mutate`.

File: extensions/amp-form/0.1/amp-form.js [hunk 16]
```diff
@@ -176,9 +188,75 @@ export class AmpForm {
...
+  input.addEventListener('keyup', onInvalidInputKeyUp_);
+  input.addEventListener('blur', onInvalidInputBlur_);
+  validationBubble.show(input, input.validationMessage);
+}
+
+
 /**
  * Installs submission handler on all forms in the document.
  * @param {!Window} win
+ * @private
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: I don't think there's a need to mark un-exported functions as private.

File: extensions/amp-form/0.1/amp-form.js [hunk 17]
```diff
@@ -189,10 +267,17 @@ function installSubmissionHandlers(win) {
...
+/**
+ * @param {!Window} win
+ * @private
+ */
 function installAmpForm(win) {
   return getService(win, 'amp-form', () => {
     if (isExperimentOn(win, TAG)) {
-      installStyles(win.document, CSS, () => installSubmissionHandlers(win));
+      installStyles(win.document, CSS, () => {
+        validationBubble = new ValidationBubble(window);
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: I'm not sure I like having a global `validationBubble`. Why not just make it an instance variable on `AmpForm`? What if there are multiple forms on the page?

File: extensions/amp-form/0.1/validation-bubble.js [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    this.bubbleElement_.classList.add('-amp-validation-bubble');
+    win.document.body.appendChild(this.bubbleElement_);
+  }
+
+  /**
+   * Hides the bubble off screen.
+   */
+  hide() {
+    this.vsync_.run({
+      mutate: hideBubble,
```
Comment (jridgewell) [inline extensions/amp-form/0.1/validation-bubble.js]: Ditto `{measure, mutate}` hidden class.

File: extensions/amp-form/0.1/amp-form.js (hunk 16 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Yeah that makes sense, @erwinmombay @cramforce shoild we not use `@private` at all when defining file-level methods/vars?

File: extensions/amp-form/0.1/amp-form.js (hunk 17 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: We really only need one validation bubble. Even with multiple forms, we only need one because we're always showing the bubble on one field at a time. I don't really see reasons to have multiple instances here.

File: extensions/amp-form/0.1/amp-form.js (hunk 16 above)
Comment (cramforce) [inline extensions/amp-form/0.1/amp-form.js]: +1

File: extensions/amp-form/0.1/amp-form.js (hunk 15 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Will do. This is for type matching of the `Task` def, right?

Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Yup.

File: extensions/amp-form/0.1/amp-form.js [hunk 19]
```diff
@@ -176,9 +188,75 @@ export class AmpForm {
...
+}
+
+
+/**
+ * Revalidates the currently focused input after a change.
+ * @param {!KeyboardEvent} event
+ * @private
+ */
+function onInvalidInputKeyUp_(event) {
+  validationBubble.hide();
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: This will schedule two `VSync` tasks, when only one is ever necessary. Either it's valid, and we can hide it, or it's invalid, and we need to update and center it.

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: It's always `element.querySelectorAll('.class')` or such. The trick is that adding `:scope` to it is a lot faster. The problem: `:scope` is not available everywhere so we can't do it universally. An inconvenience for a generic solution is that you have to have `:scope` for each sub-selector. I.e. `input, select` becomes `:scope input, :scope select`.

Alternatively, you can use `element.getElementsByTagName()` API or such: obviously given three selectors this might not be better.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Sounds good. Writing up a separate PR for the `dom.js` changes. Thanks for the explanation

File: extensions/amp-form/0.1/amp-form.js [hunk 20]
```diff
@@ -111,7 +111,10 @@ export class AmpForm {
     if (shouldValidate &&
         this.form_.checkValidity && !this.form_.checkValidity()) {
       e.preventDefault();
-      this.vsync_.run({mutate: reportValidity}, {form: this.form_});
+      this.vsync_.run({
+        measure: undefined,
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Add a TODO https://github.com/ampproject/amphtml/issues/3776, please.

File: extensions/amp-form/0.1/amp-form.js [hunk 21]
```diff
@@ -177,6 +192,73 @@ export class AmpForm {
...
+}
+
+
+/**
+ * Revalidates the currently focused input after a change.
+ * @param {!KeyboardEvent} event
+ */
+function onInvalidInputKeyUp_(event) {
+  if (event.target.checkValidity()) {
+    validationBubble.hide();
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: One last: We may end up double calling `#hide` and `#show`. Storing the current message as a state on `ValidationBubble`, we can check if we're already hidden, or if we're trying to show the same message again. In either case, we can return without doing anything.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Yeah we'd also need to store the `targetElement` in that state to make sure we're calling show for the same element as well.

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Sent out a PR #3778

File: extensions/amp-form/0.1/validation-bubble.js [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+
+/**
+ * Measures the layout for the target element passed through state object.
+ * @param {!Object} state
+ * @private
+ */
+function measureTargetElement(state) {
+  state.targetRect = state.viewport.getLayoutRect(state.targetElement);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/validation-bubble.js]: Will it ever calculate with `display: none`?

File: extensions/amp-form/0.1/validation-bubble.js [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * Updates text content, positions and displays the bubble.
+ * @param {!Object} state
+ * @private
+ */
+function showBubbleElement(state) {
+  state.bubbleElement.textContent = state.message;
+  setStyles(state.bubbleElement, {
+    display: 'initial',
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/validation-bubble.js]: Is `display: initial` supported by all platforms we need? It might be better to just set it to `''`

File: extensions/amp-form/0.1/validation-bubble.js (hunk 22 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: `targetElement` is not `display:none` that is the `bubbleElement`.

File: extensions/amp-form/0.1/amp-form.js (hunk 21 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Filed a bug to track this will do in another PR #3782

File: extensions/amp-form/0.1/amp-form.js (hunk 15 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 16 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/validation-bubble.js (hunk 18 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 19 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 20 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/validation-bubble.js (hunk 23 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/validation-bubble.js]: Done.

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: As discussed in #3778 this isn't necessary.

**Comment by mkhatib** [comment]
Added tests and addressed all comments. PTAL.

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (erwinmombay) [inline extensions/amp-form/0.1/amp-form.js]: very anecdotal but :scope has always been faster for me because querySelectorAll is not element rooted in the way it works. (it does not work the way you think it does)

see http://ejohn.org/blog/thoughts-on-queryselectorall/ for explanation.

**Comment by jridgewell** [comment]
LGTM.
